Content.Server/Radiation/Systems/RadiationSystem.GridCast.cs

```diff
@@ -0,0 +1,239 @@
+using System.Linq;
+using Content.Server.Radiation.Components;
+using Content.Shared.Physics;
+using Content.Shared.Radiation.Components;
+using Content.Shared.Radiation.Events;
+using Content.Shared.Radiation.Systems;
+using Robust.Shared.Map;
+using Robust.Shared.Physics;
+using Robust.Shared.Timing;
+
+namespace Content.Server.Radiation.Systems;
+
+// main algorithm that fire radiation rays to target
+public partial class RadiationSystem
+{
+    private void UpdateGridcast()
+    {
+        var stopwatch = new Stopwatch();
+        stopwatch.Start();
+
+        var sources = EntityQuery<RadiationSourceComponent, TransformComponent>().ToArray();
+        var destinations = EntityQuery<RadiationReceiverComponent, TransformComponent>().ToArray();
+        var blockerQuery = GetEntityQuery<RadiationBlockerComponent>();
+        var resistanceQuery = GetEntityQuery<RadiationGridResistanceComponent>();
+
+        // trace all rays from rad source to rad receivers
+        var rays = new List<RadiationRay>();
+        foreach (var (source, sourceTrs) in sources)
+        {
```

Comment:
The source world position should be computed in the outer for loop and re-used, instead of recomputing for each receiver. 

Also, using the `SharedTransformSystem.GetWorldPosition()` override that takes in a transform query to compute the world positions will be faster than using the `xform.WorldPosition` getter